README.md
# crawler-request-js
Forked from https://gitlab.com/autokent/crawler-request.
Modified to pass metadata back from HTML, PDF, and Open Office documents.

**HTTP request module customized for crawlers.**

## Usage

### Simple Request
```js
const crawler = require('crawler-request-js');

crawler("https://stackoverflow.com/").then(function(response){
    // handle response
    console.log(response.text.length);
});
```

### PDF Parse
```js
const crawler = require('crawler-request-js');

crawler("http://careers.stackoverflow.com/stack_overflow_careers.pdf").then(function(response){
    // handle response
    console.log(response.text.length);
});
```

### HTML Metadata
```js
const crawler = require('crawler-request-js');

crawler("http://stackoverflow.com/").then(function(response){
    // handle response
    console.log(response.metadata.Title);
    console.log(response.metadata.Author);
    console.log(response.metadata.ModDate);
});
```

### PDF Metadata
```js
const crawler = require('crawler-request-js');

crawler("http://careers.stackoverflow.com/stack_overflow_careers.pdf").then(function(response){
    // handle response
    console.log(response.metadata['dc:title'];
    console.log(response.metadata['dc:creator']);
    console.log(response.metadata['dc:description']);
    console.log(response.metadata['xmp:modifydate']);
});
```

### Open Office Metadata
```js
const crawler = require('crawler-request-js');

crawler("http://example.com/spreadsheet.xlsx").then(function(response){
    // handle response
    console.log(response.metadata.editable.title.value);
    console.log(response.metadata.editable.creator.value);
    console.log(response.metadata.editable.modified.value);
});
```

### Extend
```js
const crawler = require('crawler-request-js');

function response_text_size(response){
    response["size"] = response.text.length;
    return response;
}

crawler("https://stackoverflow.com/",response_text_size).then(function(response){
    // handle response
	console.log(response.size);
});
```

## License
[MIT licensed](https://github.com/stewarjs/crawler-request/blob/master/LICENSE) and all it's dependencies are MIT or BSD licensed.


CHANGELOG
# 1.3.0
* forked from https://gitlab.com/autokent/crawler-request
* return HTML and PDF metadata with response

# 1.2.2
* co dependency removed.

# 1.2.0 - 1.2.1
* files organized.

# 1.1.9
* pdf parsing issue fixed.


# 1.1.3 - 1.1.8
* files organized.

# 1.1.1 & 1.1.2
* ci token security issue fixed

# 1.1.0
* initial release

test/test.js
const assert = require('assert');
const cr = require('../');

describe('negative-positive must be equal', function() {
    it('should equal', function() {
    
    });
});




test/general.js
const assert = require('assert');
const CR = require('../');

describe('must crawl google', function () {
	this.timeout("1200000");
	it('should set error when url is invalid', function () {
		let url = "aaabbbccc";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "none");
			assert.equal(res.status, -1);
			assert.notEqual(res.error, null);
		});
	});

	it('should set error when url has invalid extension', function () {
		let url = "https://stackoverflow.com/logo.png";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "none");
			assert.equal(res.status, -1);
			assert.notEqual(res.error, null);
			assert.equal(res.error, "unsupported-extension");
		})
	});


	it('should return response when url has http protocol-1', function () {
		let url = "http://stackoverflow.com";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "html");
			assert.equal(res.status, 200);
			assert.equal(res.error, null);
			assert.notEqual(res.text, null);
			assert.notEqual(res.text.length, 0);
		})
	});

	it('should return response when url has http protocol-2', function () {
		let url = "http://stackoverflow.com/";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "html");
			assert.equal(res.status, 200);
			assert.equal(res.error, null);
			assert.notEqual(res.text, null);
			assert.notEqual(res.text.length, 0);
		})
	});

	it('should return response when url has https protocol-1', function () {
		let url = "https://stackoverflow.com";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "html");
			assert.equal(res.status, 200);
			assert.equal(res.error, null);
			assert.notEqual(res.text, null);
			assert.notEqual(res.text.length, 0);
		})
	});

	it('should return response when url has https protocol-2', function () {
		let url = "https://stackoverflow.com/";
		return CR(url).then(function (res) {
			assert.equal(res.url, url);
			assert.equal(res.type, "html");
			assert.equal(res.status, 200);
			assert.equal(res.error, null);
			assert.notEqual(res.text, null);
			assert.notEqual(res.text.length, 0);
		})
	});

});


test/callback-test.js
const assert = require('assert');
const CR = require('../');

describe('must callbacks works', function() {
    this.timeout("60000");

    it('should single callback works', function(done) {

        function response_text_size(response){
            response["size"] = response.text.length;
            return response;
        }

        CR("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf",response_text_size).then(function(response){
            //handle response
            assert.notEqual(response.size,null);
            assert.notEqual(response.size,0);
            done();
        });

    });
});

QUICKSTART.js
//const CR = require('crawler-request');
const CR = require('./');
const Co = require('co');
const Cheerio = require('cheerio');
const Fs = require("fs");
const Url = require('url');
const callbackArr = [];

function test1(result) {
    result["test1"] = "test1";
    return result;
}

function test2(result) {
    result["test2"] = "test2";
    return result;
}

function text_size(result) {
    result["text_size"] = result.text == null ? 0 : result.text.length;
    return result;
}

callbackArr.push(test1);
callbackArr.push(test2);
callbackArr.push(text_size);

var urlArr = [];
urlArr.push("http://www.mu.edu.tr/tr/personel/osaygin");
urlArr.push("http://www.saglikbilimleri.mu.edu.tr/tr/beden-egitimi-ve-spor-anabilim-dali-akademik-kurulu-uyeleri-5461");
urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/2010/815");
urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/843/406");
urlArr.push("http://dergipark.gov.tr/inubesyo/board");
urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
urlArr.push("http://www.acarindex.com/dosyalar/makale/acarindex-1423936666.pdf");
urlArr.push("https://www.cabdirect.org/cabdirect/abstract/20093291780?start=2050");
urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
urlArr.push("http://www.eng.utah.edu/~lzang/images/Lecture_10_AFM.pdf");
urlArr.push("http://www.groupes.polymtl.ca/jnmes/archives/article_in_press/JNMES-1029.pdf");
urlArr.push("https://www.researchgate.net/publication/315951188_Atomic_force_microscopy_investigation_of_step_generation_and_bunching_on_100_facet_of_CH_3_NH_3_PbI_3_crystal_grown_from_g-Butyrolactone_Atomic_force_microscopy_investigation_of_step_generation_and_bu");
urlArr.push("http://ezfind.technion.ac.il/vufind/EDS/Search?lookfor=SINGLE+crystals&type=SU");
urlArr.push("http://onlinelibrary.wiley.com/doi/10.1002/crat.201700021/abstract");


// test code starts here

let testing_purpose = Co.wrap(function* () {
    
    let response_01 = yield CR("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all");
    debugger;

    let response_02 = yield CR("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all",callbackArr);
    debugger;

    let response_03 = yield CR("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
    debugger;

    let response_04 = yield CR("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf", callbackArr);
    debugger;

    let response_05 = yield CR("http://daytam.atauni.edu.tr/uploads/AFM.docx", callbackArr);
    debugger;

    let response_06 = yield CR("aaabbbccc");
    debugger;

    let response_07 = yield CR("aaabbbccc", callbackArr);
    debugger;

    let response_08 = yield CR(urlArr, callbackArr);
    debugger;
});
    
testing_purpose();






crawler-request.js
const Axios = require('axios');
const FileType = require('file-type');
const ContentType = require('content-type');
const IconvLite = require('iconv-lite');
const HtmlToText = require('html-to-text');
const cheerio = require('cheerio');
const PdfParse = require('pdf-parse');


const html_to_text_options = {
	tables: true,
	wordwrap: false,
	ignoreImage: true,
	preserveNewlines: false,
	uppercaseHeadings: false,
	hideLinkHrefIfSameAsText: true,
	baseElement: ['html'],
	ignoreHref: true,
	//wordwrap:150,
	noLinkBrackets: true
};

const MIME_REGEX = /.*\.(jpg|png|gif|dotx|doc|webp|flif|cr2|tif|bmp|jxr|psd|rar|zip|tar|rar|js|gz|bz2|7z|dmg|mp4|m4v|mid|mkv|webm|mov|avi|wmv|mpg|mp3|m4a|ogg|opus|flac|wav|amr|epub|exe|swf|rtf|woff|woff2|eot|ttf|otf|ico|flv|ps|xz|sqlite|nes|crx|xpi|cab|dep|ar|rpm|z|lz|msi|mxf|mts|wasm|blend|bpg|docx|pptx|xlsx|3gp|css|xlam|xla|xls|xps|exe)$/i;

function _crawler_request(current_url) {
	let instance = Axios.create();

	let request_config = {
		url: current_url,
		method: 'get',
		baseURL: '',

		transformResponse: [function (data, headers) {

			let ret = {
				url: current_url,
				type: "none",
				html: null,
				text: null,
				metadata: array(),
				status: null,
				error: null
			};

			let type_str = headers['content-type'];

			if (type_str) {

				if (type_str.match(/\/(x-)?pdf/ig)) {
					ret.type = "pdf";
				} else if (type_str.match(/\/html/ig) || type_str.match(/\/json/ig)) {

					let charset = 'none';
					let type_obj = ContentType.parse(type_str);

					if (typeof type_obj.parameters['charset'] !== 'undefined') {
						charset = type_obj.parameters['charset'].toLowerCase();
					}

					if (charset == 'utf-8' || charset == 'ascii' || charset == 'none' || !IconvLite.encodingExists(charset)) {
						var $ = cheerio.load(data.toString());
						ret.metadata = [{
							title: $('head > title').text(),
							author: $('meta[name="author"]').attr('content')
						}];
						ret.html = data.toString();
						ret.type = "html";
					} else {
						ret.html = IconvLite.decode(data, charset);
						ret.type = "html";
					}
				} else {
					ret.type = "none";
				}
			}

			let mimeType = FileType(data);

			if (mimeType == null) {
				ret.html = data.toString();
				ret.type = "html";
			} else if (mimeType.ext == 'pdf') {
				ret.type = "pdf";
			} else if (mimeType.mime != null && mimeType.mime.match(/\/(x-)?pdf/ig)) {
				ret.type = "pdf";
			} else {
				//not supported mime-type
				ret.type = "none";

			}

			//at this point file type is dedected.



			if (ret.type == "html") {
				//ret.text = HtmlToText.fromString(ret.html, html_to_text_options);
				return ret;
			} else if (ret.type == "pdf") {
				return PdfParse(data)
					.then(res => {
						if (res) {
							ret.metadata = res.metadata;
							ret.text = res.text;
						}

						ret.type = "pdf";
						return ret;
					})
					.catch(err => {
						ret.status = -222;
						ret.error = err.toString();
						return ret;
					});
			} else {
				return ret;
			}
		}],

		headers: {
			'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36'
		},

		timeout: 40000,
		withCredentials: false,
		responseType: 'arraybuffer', //'arraybuffer', 'blob', 'document', 'json', 'text', 'stream'
		maxContentLength: 20000000,
		validateStatus: function (status) {
			return status >= 200 && status < 300;
		},
		maxRedirects: 5,
	};

	let current_status = -100;

	if (MIME_REGEX.test(current_url)) {
		return Promise.resolve({
			url: current_url,
			type: "none",
			html: null,
			text: null,
			metadata: array(),
			status: -100,
			error: "unsupported-extension"
		});
	}

	return instance.request(request_config)
		.then(function (res) {
			current_status = res.status;
			return res.data;
		})
		.then(function (res) {
			res.status = res.status == -222 ? -222 : current_status;
			return res;
		})
		.catch(function (err) {
			return {
				url: current_url,
				type: "none",
				html: null,
				text: null, //err.response.status
				metadata: array(),
				status: err.response && err.response.status ? err.response.status : -111,
				error: err.toString()
			};
		});
}


function crawler_request_wrapper(current_url, callbackArr) {
	if (typeof current_url != "string" && !Array.isArray(current_url)) throw "first parameter must be a string or an array.";
	if (typeof callbackArr != "undefined" && typeof callbackArr != "function" && !(callbackArr instanceof Array)) throw "second parameter must be a array or callback.";
	if (typeof callbackArr == "undefined") callbackArr = [];
	if (!Array.isArray(callbackArr)) callbackArr = [callbackArr];


	if (Array.isArray(current_url)) {
		let promiseArr = [];

		for (let url of current_url) {
			let crawler_promise = _crawler_request(url);

			for (let callback of callbackArr) {
				crawler_promise = crawler_promise.then(function (result) {
					return callback(result);
				});
			}
			promiseArr.push(crawler_promise);
		}

		return Promise.all(promiseArr);
	} else {

		let crawler_promise = _crawler_request(current_url);

		for (let callback of callbackArr) {
			crawler_promise = crawler_promise.then(function (result) {
				return callback(result);
			});
		}

		return crawler_promise;
	}
}

module.exports = crawler_request_wrapper;

//for testing purpose
if (!module.parent) {
	let Cheerio = require('cheerio');
	let Fs = require("fs");
	let Url = require('url');
	let callbackArr = [];

	function test1(result) {
		result["test1"] = "test1";
		return result;
	}

	function test2(result) {
		result["test2"] = "test2";
		return result;
	}

	function text_size(result) {
		result["text_size"] = result.text == null ? 0 : result.text.length;
		return result;
	}

	callbackArr.push(test1);
	callbackArr.push(test2);
	callbackArr.push(text_size);

	var urlArr = [];
	urlArr.push("http://www.mu.edu.tr/tr/personel/osaygin");
	urlArr.push("http://www.saglikbilimleri.mu.edu.tr/tr/beden-egitimi-ve-spor-anabilim-dali-akademik-kurulu-uyeleri-5461");
	urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/2010/815");
	urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/843/406");
	urlArr.push("http://dergipark.gov.tr/inubesyo/board");
	urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
	urlArr.push("http://www.acarindex.com/dosyalar/makale/acarindex-1423936666.pdf");
	urlArr.push("https://www.cabdirect.org/cabdirect/abstract/20093291780?start=2050");
	urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
	urlArr.push("http://www.eng.utah.edu/~lzang/images/Lecture_10_AFM.pdf");
	urlArr.push("http://www.groupes.polymtl.ca/jnmes/archives/article_in_press/JNMES-1029.pdf");
	urlArr.push("https://www.researchgate.net/publication/315951188_Atomic_force_microscopy_investigation_of_step_generation_and_bunching_on_100_facet_of_CH_3_NH_3_PbI_3_crystal_grown_from_g-Butyrolactone_Atomic_force_microscopy_investigation_of_step_generation_and_bu");
	urlArr.push("http://ezfind.technion.ac.il/vufind/EDS/Search?lookfor=SINGLE+crystals&type=SU");
	urlArr.push("http://onlinelibrary.wiley.com/doi/10.1002/crat.201700021/abstract");


	async function testing_purpose() {

		//let result_01 = await crawler_request_wrapper("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all");
		//debugger;

		//let result_02 = await crawler_request_wrapper("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all",callbackArr);
		//debugger;

		//let result_03 = await crawler_request_wrapper("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
		//debugger;

		//let result_04 = await crawler_request_wrapper("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf", callbackArr);
		//debugger;

		//let result_05 = await crawler_request_wrapper("http://daytam.atauni.edu.tr/uploads/AFM.docx", callbackArr);
		//debugger;

		//let result_06 = await crawler_request_wrapper("aaabbbccc");
		//debugger;

		//let result_07 = await crawler_request_wrapper("aaabbbccc", callbackArr);
		//debugger;

		//let result_08 = await crawler_request_wrapper(urlArr, callbackArr);
		//debugger;


		//while (true) {
		//	let result_10 = await crawler_request_wrapper("http://www.fizik.itu.edu.tr/physics-10x/doc/FIZ101E_2014-2017.rar");
		//	if (result_10.error != "unsupported-extension") debugger;
		//}


		//let result_11 = await crawler_request_wrapper("https://www.nanomagnetics-inst.com/usrfiles/files/Articles/RT-SHPM/RT-SHPM-1.pdf");
		//debugger;

		let result_12 = await crawler_request_wrapper("https://www.nanomagnetics-inst.com/usrfiles/files/Flyers/N_Akin_makale-2015.pdf");
		debugger;




		//process.exit();

	};

	testing_purpose();
}


crawler-request-js.js
const Axios = require('axios');
const FileType = require('file-type');
const ContentType = require('content-type');
const IconvLite = require('iconv-lite');
const cheerio = require('cheerio');
const PdfParse = require('pdf-parse');
const OFFICEPROPS = require('officeprops');


const PARSE_PDF_OPTIONS = {
	// max page number to parse
	max: 1
}


// Skip these extensions
const MIME_REGEX = /.*\.(jpg|png|gif|dotx|webp|flif|cr2|tif|bmp|jxr|psd|rar|zip|tar|rar|js|gz|bz2|7z|dmg|mp4|m4v|mid|mkv|webm|mov|avi|wmv|mpg|mp3|m4a|ogg|opus|flac|wav|amr|epub|exe|swf|rtf|woff|woff2|eot|ttf|otf|ico|flv|ps|xz|sqlite|nes|crx|xpi|cab|dep|ar|rpm|z|lz|msi|mxf|mts|wasm|blend|bpg|3gp|css|xlam|xla|xps|exe)$/i;

function _crawler_request(current_url) {
	let instance = Axios.create();

	let request_config = {
		url: current_url,
		method: 'get',
		baseURL: '',

		transformResponse: [function (data, headers) {

			let ret = {
				url: current_url,
				type: "none",
				html: null,
				text: null,
				metadata: null,
				status: null,
				error: null
			};
            
            // determine file type

			let type_str = headers['content-type'];

			if (type_str) {

				if (type_str.match(/\/(x-)?pdf/ig)) {
					ret.type = "pdf";
				} else if (type_str.match(/\/html/ig) || type_str.match(/\/json/ig)) {

					let charset = 'none';
					let type_obj = ContentType.parse(type_str);

					if (typeof type_obj.parameters['charset'] !== 'undefined') {
						charset = type_obj.parameters['charset'].toLowerCase();
					}

					if (charset == 'utf-8' || charset == 'ascii' || charset == 'none' || !IconvLite.encodingExists(charset)) {
						ret.html = data.toString();
						ret.type = "html";
					} else {
						ret.html = IconvLite.decode(data, charset);
						ret.type = "html";
					}
				} else if (type_str.match(/\/vnd.openxmlformats-officedocument.presentationml.presentation/ig) || type_str.match(/\/vnd.ms-powerpoint/ig)) {
					ret.type = "ppt";
				} else if (type_str.match(/\/vnd.openxmlformats-officedocument.wordprocessingml.document/ig) || type_str.match(/\/msword/ig)) {
					ret.type = "doc";
				} else if (type_str.match(/\/vnd.openxmlformats-officedocument.spreadsheetml.sheet/ig) || type_str.match(/\/vnd.ms-excel/ig)) {
					ret.type = "xls";
				} else {
					ret.type = type_str;
				}
			}


            // get metadata and plaintext
			if (ret.type == "html") {
                var $ = cheerio.load(ret.html);
                ret.metadata = {
                    Title: $('title').text(),
                    Author: $('meta[name="author"]').attr('content'),
                    Description: $('meta[name="description"]').attr('content'),
                    ModDate: $('meta[http-equiv="last-modified"]').attr('content')
                };
				ret.text = $('#main').text().replace( /\r?\n|\r/g, " " );
				ret.html = null;
				return ret;
			} else if (ret.type == "pdf") {
				return PdfParse(data, PARSE_PDF_OPTIONS)
					.then(res => {
						if (res) {
							//ret.metadata = res.metadata._metadata;
                            ret.metadata = {
                                Title: res.metadata._metadata['dc:title'],
                                Author: res.metadata._metadata['dc:creator'],
                                Description: res.metadata._metadata['dc:description'],
                                ModDate: res.metadata._metadata['xmp:modifydate']
                            };
							ret.text = res.text.replace( /\r?\n|\r/g, " " );
						}

						return ret;
					})
					.catch(err => {
						ret.status = -222;
						ret.error = err.toString();
						return ret;
					});
			} else if (ret.type == "doc" || ret.type == "ppt" || ret.type == "xls") {
				return OFFICEPROPS.getData(data).then(function(metadata){
                    //ret.metadata = metadata;
                    ret.metadata = {
                        Title: metadata.editable.title.value,
                        Author: metadata.editable.creator.value,
                        Description: metadata.editable.description.value,
                        ModDate: metadata.editable.modified.value
                    };
                    return ret;
                }).catch(err => {
						ret.status = -222;
						ret.error = err.toString();
						return ret;
					});
			} else {
				return ret;
			}
		}],

		headers: {
			'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36'
		},

		timeout: 40000,
		withCredentials: false,
		responseType: 'arraybuffer', //'arraybuffer', 'blob', 'document', 'json', 'text', 'stream'
		maxContentLength: 20000000,
		validateStatus: function (status) {
			return status >= 200 && status < 300;
		},
		maxRedirects: 5,
	};

	let current_status = -100;

	if (MIME_REGEX.test(current_url)) {
		return Promise.resolve({
			url: current_url,
			type: "none",
			html: null,
			text: null,
			metadata: null,
			status: -100,
			error: "unsupported-extension"
		});
	}

	return instance.request(request_config)
		.then(function (res) {
			current_status = res.status;
			return res.data;
		})
		.then(function (res) {
			res.status = res.status == -222 ? -222 : current_status;
			return res;
		})
		.catch(function (err) {
			return {
				url: current_url,
				type: "none",
				html: null,
				text: null, //err.response.status
				metadata: null,
				status: err.response && err.response.status ? err.response.status : -111,
				error: err.toString()
			};
		});
}


function crawler_request_wrapper(current_url, callbackArr) {
	if (typeof current_url != "string" && !Array.isArray(current_url)) throw "first parameter must be a string or an array.";
	if (typeof callbackArr != "undefined" && typeof callbackArr != "function" && !(callbackArr instanceof Array)) throw "second parameter must be a array or callback.";
	if (typeof callbackArr == "undefined") callbackArr = [];
	if (!Array.isArray(callbackArr)) callbackArr = [callbackArr];


	if (Array.isArray(current_url)) {
		let promiseArr = [];

		for (let url of current_url) {
			let crawler_promise = _crawler_request(url);

			for (let callback of callbackArr) {
				crawler_promise = crawler_promise.then(function (result) {
					return callback(result);
				});
			}
			promiseArr.push(crawler_promise);
		}

		return Promise.all(promiseArr);
	} else {

		let crawler_promise = _crawler_request(current_url);

		for (let callback of callbackArr) {
			crawler_promise = crawler_promise.then(function (result) {
				return callback(result);
			});
		}

		return crawler_promise;
	}
}

module.exports = crawler_request_wrapper;

//for testing purpose
if (!module.parent) {
	let Cheerio = require('cheerio');
	let Fs = require("fs");
	let Url = require('url');
	let callbackArr = [];

	function test1(result) {
		result["test1"] = "test1";
		return result;
	}

	function test2(result) {
		result["test2"] = "test2";
		return result;
	}

	function text_size(result) {
		result["text_size"] = result.text == null ? 0 : result.text.length;
		return result;
	}

	callbackArr.push(test1);
	callbackArr.push(test2);
	callbackArr.push(text_size);

	var urlArr = [];
	/*urlArr.push("http://www.mu.edu.tr/tr/personel/osaygin");
	urlArr.push("http://www.saglikbilimleri.mu.edu.tr/tr/beden-egitimi-ve-spor-anabilim-dali-akademik-kurulu-uyeleri-5461");
	urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/2010/815");
	urlArr.push("https://www.j-humansciences.com/ojs/index.php/IJHS/article/viewFile/843/406");
	urlArr.push("http://dergipark.gov.tr/inubesyo/board");
	urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
	urlArr.push("http://www.acarindex.com/dosyalar/makale/acarindex-1423936666.pdf");
	urlArr.push("https://www.cabdirect.org/cabdirect/abstract/20093291780?start=2050");
	urlArr.push("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
	urlArr.push("http://www.eng.utah.edu/~lzang/images/Lecture_10_AFM.pdf");
	urlArr.push("http://www.groupes.polymtl.ca/jnmes/archives/article_in_press/JNMES-1029.pdf");
	urlArr.push("https://www.researchgate.net/publication/315951188_Atomic_force_microscopy_investigation_of_step_generation_and_bunching_on_100_facet_of_CH_3_NH_3_PbI_3_crystal_grown_from_g-Butyrolactone_Atomic_force_microscopy_investigation_of_step_generation_and_bu");
	urlArr.push("http://ezfind.technion.ac.il/vufind/EDS/Search?lookfor=SINGLE+crystals&type=SU");
	urlArr.push("http://onlinelibrary.wiley.com/doi/10.1002/crat.201700021/abstract");*/


	async function testing_purpose() {

		//let result_01 = await crawler_request_wrapper("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all");
		//debugger;

		//let result_02 = await crawler_request_wrapper("https://stackoverflow.com/questions/42772401/parallel-operations-with-promise-all",callbackArr);
		//debugger;

		//let result_03 = await crawler_request_wrapper("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf");
		//debugger;

		//let result_04 = await crawler_request_wrapper("http://journals.tubitak.gov.tr/medical/issues/sag-09-39-3/sag-39-3-4-0902-21.pdf", callbackArr);
		//debugger;

		//let result_05 = await crawler_request_wrapper("http://daytam.atauni.edu.tr/uploads/AFM.docx", callbackArr);
		//debugger;

		//let result_06 = await crawler_request_wrapper("aaabbbccc");
		//debugger;

		//let result_07 = await crawler_request_wrapper("aaabbbccc", callbackArr);
		//debugger;

		//let result_08 = await crawler_request_wrapper(urlArr, callbackArr);
		//debugger;


		//while (true) {
		//	let result_10 = await crawler_request_wrapper("http://www.fizik.itu.edu.tr/physics-10x/doc/FIZ101E_2014-2017.rar");
		//	if (result_10.error != "unsupported-extension") debugger;
		//}


		//let result_11 = await crawler_request_wrapper("https://www.nanomagnetics-inst.com/usrfiles/files/Articles/RT-SHPM/RT-SHPM-1.pdf");
		//debugger;

		/*let result_12 = await crawler_request_wrapper("https://www.nanomagnetics-inst.com/usrfiles/files/Flyers/N_Akin_makale-2015.pdf");
		debugger;*/




		//process.exit();

	};

	testing_purpose();
}
